### ISSUE TITLE
Security of 2FA enrollment process may be improved.

### ISSUE THREAD

**Issue by n4kre** [authelia/authelia#135]
Hi,

Let's assume the following scenario to illustrate the risk refered by this issue's title and explain how this may ruin the security benefit of 2FA.

- An attacker manages to steal valid credentials.
- The victim uses the same password to access his e-mail service.
- The attacker uses these credentials access the victim's e-mails.
- The attacker can enroll his own device as a 2nd factor to Authelia.
- Victim's data protected by Authelia can be accessed and compromised by the attacker.

As a result, the whole benefit of 2FA is defeated – assuming the user had the same password for both his Authelia account and e-mail service, and yes, password reuse is unfortunately very frequent. Therefore, I would advice the following behavior to lower such a risk.

1. A user wants to register a second factor.
2. An e-mail is sent to the admin  – yes… another parameter in Authelia's configuration file –, which may approve or not this request. Such an e-mail may look like this:

>`username` asked on `date` to enroll a new device as a second factor (`TOTP` | `U2F`) for its account on `login.example.com`.
>
>This request was sent from `IP` (`Country`, `Region`, `City` | unkown location *[if geo-location has failed]*).
>
>Do you approve this request?
>
>[a green “Yes” button] [a red “No” button]

3. If the admin approves this request, the user receives the e-mail he would receives in the current Authelia behavior, to actually register a second factor.

Step 2. makes the aformentioned attack significantly more difficult, since the admin approval is required. The idea is that it may get in touch – physically, or by phone – with the user to make sure such a request was legitimate, in case it was unlikely to occur – weird IP location, user had already enrolled and has not reported loss or theft of its phone/YubiKey, etc.

**Note**: In step 3., keep it as it currently is: it is very neat – you do not ever send any sensitive information through e-mail, evey link sent by Authelia is valid only one time and for a limited duration. 👍  If you decide to use IP location, using some kind of geo-location package, note that it wouldn't hurt to always include these details for every e-mail that is sent by Authelia: it helps to detect malicious activity.

**Comment by clems4ever** [comment]
You are right, this scenario is clearly plausible. Thank you very much for the detailed description of the use case. 
And yes, the IP and geo-location is a good idea to put in the email anyway. I don't know how to get the geo-location, though. I will have a look.
Thank you very much.

**Comment by n4kre** [comment]
There are various “geoip” packages. Maybe [this one](https://github.com/runk/node-maxmind) would be suitable.

Thank you!

**Comment by jmccoy555** [comment]
Hello all. 
It's great to see such active development here, thanks to all. 
My use case for Authelia is to actually beef up the security of my webmail (SOGo) and other web apps when accessing from an external network. So it would be good to keep this authorisation mechanism optional, or only implement when accessed from external or a certain list / or not of IP networks? 
Actually, not wanting to hijack this, and maybe I should open a separate feature request, but I was wondering if it is / will be possible to pass through the user credentials for true SSO with back end applications, although I think this may require Authelia to support SAML?

**Comment by clems4ever** [comment]
Hello @jmccoy555 , I'm in the same case as you. I'm securing some web apps and I'm pretty much the only user accessing those services. Thus, such a mechanism is a bit overkill for my use case indeed. But I clearly see the point of @n4kre when you are in an organization when you don't control everything. I will try to find a good way to implement this feature and a way to opt-in the feature.

Concerning your second request, please create a new ticket.

**Comment by jmccoy555** [comment]
Hi @clems4ever , I can certainly appreciate @n4kre use case, and his insight is particularly interesting for someone like me that with little knowledge on this subject; and is certainly making me think!! 

Thanks both.

**Comment by n4kre** [comment]
About the second factor enrollment: from my observation, since only one TOTP secret can be registered per user – I guess this holds true for U2F tokens? Anyway, this is very good not to allow multiple TOTP secrets or U2F tokens to be associated to the same user –, the behavior that I have suggested in [this post](https://github.com/clems4ever/authelia/issues/135#issue-264026347) also let the admin warn a user which wants to register an *extra* TOTP secret (or U2F token) that its *previous* TOTP secret (or U2F token) will be replaced by this one. As of now, from the user perspective, it is indeed not clear whether it can register multiple second factors.

**Comment by ciroiriarte** [comment]
It's not unusual to have a second U2F archived key in case you lose the main one..

**Comment by clems4ever** [comment]
hello @ciroiriarte , @n4kre , having multiple devices registered for a given user is a valid use case. The only thing is that it must be simple to revoke in the case one device has been lost or hijacked.

This will be handled once the client has been re-written in Angular. It will be much easier to display the currently registered devices and be able to revoke them.

**Comment by clems4ever** [comment]
Hello @jmccoy555 , 

The following use case has been extensively discussed in Gitter lately with @nightah.

> My use case for Authelia is to actually beef up the security of my webmail (SOGo) and other web apps when accessing from an external network. So it would be good to keep this authorisation mechanism optional, or only implement when accessed from external or a certain list / or not of IP networks?

We decided to implement something to handle this use case. The basic idea is that you'll be able to bind one authorization policy per network and thanks to #289 you'll be able to apply "bypass" policy for trusted networks.

**Comment by clems4ever** [comment]
Hey @jmccoy555 , there are indeed 2 topics in this thread:
1/ apply different authorization policies on different networks.
2/ Make an admin authorize the device registration request of a user.

In my last comment I talked only about 1. For 2, I need to think more about it. Having this extra layer of security costs a lot to the user who won't be able to use its new device right away.

**Comment by james-d-elliott** [comment]
This will be rolled up into #3801 where we'll overhaul areas surrounding this. We'll reopen this issue it doesn't end up being part of the final design.

**Comment by james-d-elliott** [comment]
Actually on closer inspection that 3801 doesn't cover this as I initially thought.
